types/wlr_output_power_management_v1.c

```diff
@@ -0,0 +1,247 @@
+#include <assert.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <wayland-server-core.h>
+#include <wlr/interfaces/wlr_output.h>
+#include <wlr/types/wlr_output_power_management_v1.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-output-power-management-unstable-v1-protocol.h"
+
+#define OUTPUT_POWER_MANAGER_V1_VERSION 1
+
+static void output_power_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void output_power_destroy(struct wlr_output_power_v1 *output_power) {
+	if (output_power == NULL) {
+		return;
+	}
+	wl_resource_set_user_data(output_power->resource, NULL);
+	wl_list_remove(&output_power->output_destroy_listener.link);
+	wl_list_remove(&output_power->output_enable_listener.link);
+	wl_list_remove(&output_power->link);
+	free(output_power);
+}
+
+static const struct zwlr_output_power_v1_interface output_power_impl;
+
+static struct wlr_output_power_v1 *output_power_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_output_power_v1_interface,
+		&output_power_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void output_power_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_output_power_v1 *output_power =
+		output_power_from_resource(resource);
+	output_power_destroy(output_power);
+}
+
+static void output_power_handle_output_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_output_power_v1 *output_power =
+		wl_container_of(listener, output_power, output_destroy_listener);
+	output_power_destroy(output_power);
+}
+
+static void output_power_v1_send_mode(struct wlr_output_power_v1 *output_power) {
+	enum zwlr_output_power_v1_mode mode;
+
+	mode = output_power->output->enabled ? ZWLR_OUTPUT_POWER_V1_MODE_ON :
+		ZWLR_OUTPUT_POWER_V1_MODE_OFF;
+	zwlr_output_power_v1_send_mode(output_power->resource, mode);
+}
+
+static void output_power_handle_output_enable(struct wl_listener *listener,
+		void *data) {
+	struct wlr_output_power_v1 *output_power =
+		wl_container_of(listener, output_power, output_enable_listener);
+	output_power_v1_send_mode(output_power);
+}
+
+static void output_power_handle_set_mode(struct wl_client *client,
+		struct wl_resource *output_power_resource,
+		enum zwlr_output_power_v1_mode mode) {
+	struct wlr_output_power_v1 *output_power =
+		output_power_from_resource(output_power_resource);
+	if (output_power == NULL) {
+		return;
+	}
+
+	switch (mode) {
+	case ZWLR_OUTPUT_POWER_V1_MODE_OFF:
+	case ZWLR_OUTPUT_POWER_V1_MODE_ON:
+		break;
+	default:
+		wlr_log(WLR_ERROR, "Invalid power mode %d", mode);
+		wl_resource_post_error(output_power_resource,
+			ZWLR_OUTPUT_POWER_V1_ERROR_INVALID_MODE,
+			"Invalid power mode");
+		return;
+	}
+
+	bool ok = wlr_output_enable(output_power->output, mode);
```

Comment:
I'd prefer to have a signal for this instead of enabling/disabling the output behind the compositor's back. For instance, in Sway we have an internal field for the DPMS status, we'd need to update it there. This would also allow a compositor to decide whether or not to apply the client's request (for instance, if an output has a idle-inhibit client on it, the compositor might want to keep it on).

Reply:
@emersion so this 

a) shouldn't enable/disable the connector (and just send a signal so the compositor can do it) or 
b) should it do the enabling/disabling *and* send a signal so the compositor gets informed about it

i assume a) ?

Reply:
> shouldn't enable/disable the connector (and just send a signal so the compositor can do it)

This :P

Reply:
Clarification: we need to provide a way for the compositor not to obey to the client, so (b) won't do.